pkg/daemon/daemon.go

```diff
@@ -848,9 +872,21 @@ func (dn *Daemon) CheckStateOnBoot() error {
 
 		glog.Infof("In desired config %s", state.currentConfig.GetName())
 
+		if state.bootstrapping {
+			if err := dn.storeCurrentConfigOnDisk(state.currentConfig); err != nil {
+				return err
+			}
+		}
+
 		// All good!
 		return nil
 	}
+	if currentOnDisk != nil {
```

Comment:
I don't fully understand this 🤔 do you have a diff in mind that you can paste?

Reply:
Something like this 
```
diff --git a/pkg/daemon/daemon.go b/pkg/daemon/daemon.go
index 2f2768cc..6f9fde9c 100644
--- a/pkg/daemon/daemon.go
+++ b/pkg/daemon/daemon.go
@@ -847,11 +847,13 @@ func (dn *Daemon) CheckStateOnBoot() error {
 			return err
 		}
 	}
+	if currentOnDisk != nil && state.currentConfig.GetName() != currentOnDisk.GetName() {
+		// The on disk state (if available) is always considered truth.
+		// We want to handle the case where etcd state was restored from a backup.
+		state.currentConfig = currentOnDisk
+	}	
 
 	inDesiredConfig := state.currentConfig == state.desiredConfig
-	if currentOnDisk != nil {
-		inDesiredConfig = inDesiredConfig && state.currentConfig.GetName() == currentOnDisk.GetName()
-	}
 	if inDesiredConfig {
 		if state.pendingConfig != nil {
 			// Great, we've successfully rebooted for the desired config,
@@ -873,12 +875,6 @@ func (dn *Daemon) CheckStateOnBoot() error {
 		// All good!
 		return nil
 	}
-	if currentOnDisk != nil {
-		// reconcile what we have on disk in case something drifted
-		if state.currentConfig.GetName() != currentOnDisk.GetName() {
-			state.currentConfig = currentOnDisk
-		}
-	}
 	// currentConfig != desiredConfig, and we're not booting up into the desiredConfig.
 	// Kick off an update.
 	return dn.triggerUpdateWithMachineConfig(state.currentConfig, state.desiredConfig)
```

right?

Reply:
fixed